lib/benchee/project/benchmark.ex

```diff
@@ -0,0 +1,175 @@
+defmodule Benchee.Project.Benchmark do
+  @moduledoc """
+  Module with macro for defining project-wide benchmarks.
+  """
+  require Logger
+
+  def benchee_scenario_attr, do: :benchee_cases
+
+  @doc """
+  Module-scope setup. Optional.
+
+  Must return {:ok, state} for actual benchmark to start.
+  """
+  defmacro setup(_opts \\ [], do: doblock) do
+    quote do
+      def setup do
+        unquote(doblock)
+      end
+    end
+  end
+
+  @doc """
+  Module-scope teardown. Optional
+
+  Can return anything.
+  """
+  defmacro teardown(_opts \\ [], do: doblock) do
+    quote do
+      def teardown(state) do
+        unquote(doblock)
+      end
+    end
+  end
+
+  # credo:disable-for-next-line Credo.Check.Refactor.CyclomaticComplexity
+  defmacro benchmark(name, opts \\ [], do: {:__block__, [], exprs}) do
+    mod = __CALLER__.module
+    state_var = atom_to_var(:state)
+
+    inputs =
+      Keyword.get(opts, :inputs, []) ++
+        for {:input, _, [name, expr]} <- exprs,
+            do:
+              {name,
+               quote do
+                 unquote(block_to_fun(expr)).()
+               end}
+
+    case_args =
+      case length(inputs) do
+        0 -> []
+        _ -> [:input]
+      end
+
+    scenarios =
+      for {:scenario, _, args} <- exprs do
+        case args do
+          [name, block] ->
+            {name, block_to_fun(block, case_args)}
+
+          [name, opts, block] ->
+            {name, {block_to_fun(block, case_args), opts}}
+        end
+      end
+
+    [setup_doblock | rest] =
+      (exprs
+       |> Enum.filter(fn t -> elem(t, 0) == :setup end)
+       |> Enum.map(fn {:setup, _, [[do: doblock]]} -> doblock end)) ++
+        [
+          quote generated: true do
+            {:ok, unquote(state_var)}
+          end
+        ]
+
+    local_setup = block_to_fun(setup_doblock, [:state])
+
+    if length(rest) > 1 do
+      :ok = Logger.warn("#{inspect(mod)}: Only the first setup is used")
```

Comment:
I'd raise here instead of warn and if we can just run one setup block then I think I'd only allow it at one level... or see if we can run all setup blocks. That can also be future work and doesn't need to be there on the first attempt.

Reply:
Yes, it's interesting idea to have multiple setups, but such use cases are somewhat rare. 
Delay it for future if there'll be any need (in other words comment/issue/PR)